adam/perception/perception_frame_difference.py

```diff
@@ -0,0 +1,72 @@
+from abc import ABC
+from typing import Generic, TypeVar
+
+from attr import attrib, attrs
+from attr.validators import in_, instance_of
+from immutablecollections import ImmutableSet
+from adam.perception.developmental_primitive_perception import DevelopmentalPrimitivePerceptionFrame, ObjectPerception, \
+    PropertyPerception, RelationPerception
+
+
+class PerceptionFrameDiff(ABC):
+    r"""
+    Represents the difference between a `LanguageLearner`\ 's perception
+    of two different moments.
+    """
+
+
+@attrs(slots=True, frozen=True)
+class DevelopmentalPrimitivePerceptionFrameDiff(PerceptionFrameDiff):
+    r"""
+    Represents the difference between a `LanguageLearner`\ 's perception
+    of two different moments as sets of added and removed relations, perceived_objects, and property assertions.
+    """
+
+    added_objects: ImmutableSet["ObjectPerception"] = attrib(validator=instance_of(ImmutableSet))
```

Comment:
Prefer"
```
 added_objects: ImmutableSet["ObjectPerception"] = attrib(converter=_to_immutableset, default=immutableset())
```

This allows the user to omit this parameter if empty and also to specify it via a list, tuple, or generator.